lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,73 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_namespace(o, name, file)
+    return if file.nil?
+
+    lc_file = file.downcase
+
+    file_namespace = @file_to_namespace[lc_file]
+    if file_namespace.nil?
+      return if @file_to_namespace.key?(lc_file)
+      file_namespace = @file_to_namespace[lc_file] = namespace_for_file(lc_file)
+      return if file_namespace.nil?
+    end
+
+    if !name.downcase.start_with?(file_namespace)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+  end
+
+  # The best option from https://stackoverflow.com/questions/1723182/a-regex-that-will-never-be-matched-by-anything
+  NEVER_MATCH = '\b'.freeze
+
+  def namespace_for_file(file)
+    path = Pathname.new(file)
+
+    return nil if path.extname != ".pp"
+
+    return nil if initial_manifest(file)
+
+    path = path.each_filename.to_a
+
+    definition_dir_index = find_module_definition_dir(path)
+
+    # How can we get this result?
+    # If it is not an initial manifest, it must come from a module,
+    # and from the manifests dir there.  This may never get used...
+    return NEVER_MATCH if definition_dir_index.nil?
```

Comment:
If this was indeed used it would cause a failure since `str.start_with?(/\b/)` would give an error. To get the same effect, use a frozen string `"\b"` for NEVER_MATCH assuming no filename ever starts with a BS.

Reply:
@hlindberg , I think you are referring to line 554 above?  I think that matches your spec here.